### PR TITLE
feat: support copy into table with transform.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

support copy like

```
copy into test_table from (select (t.id+1), age from @test t)  FILE_FORMAT = (type = parquet)  force=true
```

note:

1. only projection with stage_table  is allowed in the select_stmt.
2. not allow options for stage_table,  always put them in copy_stmt.
3. must check colored files before infer_schema (so we infer on the right file), and  infer_schema is required in binding, so list_file and color_file is advanced to the binding phase, but read parquet meta is still in read_partition and list_file stage_table function is not advanced so that user can get error msg earlier instead of waiting for list_file internal.

todo:

- [x] add tests

not in pr:

-  copy  from URI

### CODE REVIEW

**Comment by youngsofun** [comment]
file seems not deleted when drop stage.

create a stage with same name may find old files there.

File: src/query/sql/src/planner/binder/copy.rs [hunk 1]
```diff
@@ -464,6 +495,148 @@ impl<'a> Binder {
         })))
     }
 
+    /// Bind COPY INFO <table> FROM <query>
```
Comment (ariesdevil) [inline src/query/sql/src/planner/binder/copy.rs:498]:
```suggestion
    /// Bind COPY INTO <table> FROM <query>
```

**Review by ariesdevil — APPROVED** [review]

File: src/query/sql/src/planner/binder/copy.rs (hunk 1 above)
Comment (youngsofun) [reply src/query/sql/src/planner/binder/copy.rs]: fixed, thanks

File: tests/suites/1_stateful/08_select_stage/08_00_parquet/08_00_06_transform.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+DATADIR_PATH="/tmp/08_00_00"
+rm -rf ${DATADIR_PATH}
+DATADIR="fs://$DATADIR_PATH/"
+echo "copy into '${DATADIR}' from t1 FILE_FORMAT = (type = PARQUET);" | $MYSQL_CLIENT_CONNECT
+
+echo "drop stage if exists s2;" | $MYSQL_CLIENT_CONNECT
+echo "create stage s2 url = '${DATADIR}' FILE_FORMAT = (type = PARQUET);"  | $MYSQL_CLIENT_CONNECT
+
+echo '--- copy 1'
+echo "copy into t1 from (select (t.id+1), age from @s2 t)  FILE_FORMAT = (type = parquet);" | $MYSQL_CLIENT_CONNECT
```
Comment (BohuTANG) [inline tests/suites/1_stateful/08_select_stage/08_00_parquet/08_00_06_transform.sh:19]: We only support parquet format, how about add an error test for:
copy into t1 from (select (t.id+1), age from @s2 t)  FILE_FORMAT = (type = csv)?

Comment (youngsofun) [reply tests/suites/1_stateful/08_select_stage/08_00_parquet/08_00_06_transform.sh]: done

File: src/query/sql/src/planner/binder/copy.rs [hunk 3]
```diff
@@ -464,6 +495,148 @@ impl<'a> Binder {
...
+
+        let start = Instant::now();
+        {
+            let status = "begin to list files";
+            self.ctx.set_status_info(status);
+            info!(status);
+        }
+
+        let operator = init_stage_operator(&stage_info)?;
+        let files = if operator.info().can_blocking() {
```
Comment (BohuTANG) [inline src/query/sql/src/planner/binder/copy.rs:544]: Not sure if the heavy IO in the planning binder is ok or not? cc @leiysky @sundy-li

Comment (youngsofun) [reply src/query/sql/src/planner/binder/copy.rs:544]: list s3

Comment (sundy-li) [reply src/query/sql/src/planner/binder/copy.rs:544]: It's already in async runtime. we don't need to use `blocking_list` I think.

Comment (youngsofun) [reply src/query/sql/src/planner/binder/copy.rs:544]: maybe faster for local fs @RinChanNOWWW ？

**Comment by youngsofun** [comment]
ydb standalone timeout

File: src/query/sql/src/planner/binder/copy.rs (hunk 3 above)
Comment (youngsofun) [reply src/query/sql/src/planner/binder/copy.rs:544]: > Not sure if the heavy IO in the planning binder is ok or not? cc @leiysky @sundy-li

no real harm for now I think.

the only other choice is delay binding and every thin to pipeline building,  which is worse.

**Comment by sundy-li** [comment]
Rest Looks good to me.

**Review by sundy-li — APPROVED** [review]

File: src/query/sql/src/planner/binder/copy.rs (hunk 3 above)
Comment (RinChanNOWWW) [reply src/query/sql/src/planner/binder/copy.rs:544]: I think `blocking_list` can let it not yield the CPU and maybe faster for local fs.

**Comment by RinChanNOWWW** [comment]
What if copy from the same stage table with different SQLs?

For example 

```sql
copy into t1 from (select t.id, age from @s2 t)
-- and
copy into t1 from (select (t.id+1), age from @s2 t)
```

Will this trigger `no file need to copy` ?

**Comment by youngsofun** [comment]
> What if copy from the same stage table with different SQLs?
> 
> For example
> 
> ```sql
> copy into t1 from (select t.id, age from @s2 t)
> -- and
> copy into t1 from (select (t.id+1), age from @s2 t)
> ```
> 
> Will this trigger `no file need to copy` ?

yes

**Comment by youngsofun** [comment]
@BohuTANG 

```
This pull request cannot be embarked for merge
The merge queue pull request can't be updated
Details:

Unable to update: user youngsofun is unknown.

Please make sure youngsofun has logged in Mergify dashboard.
```

**Pull request merged**